daemon/algod/api/server/v2/simulate.go

```diff
@@ -0,0 +1,205 @@
+	return config.Consensus[l.hdr.CurrentProtocol]
+}
+
+// GetCreatorForRound is part of LedgerForSimulator interface.
+func (l apiSimulatorLedgerConnector) GetCreatorForRound(round basics.Round, cidx basics.CreatableIndex, ctype basics.CreatableType) (creator basics.Address, ok bool, err error) {
+	if round != l.Latest() {
+		err = fmt.Errorf(
+			"GetCreatorForRound() evaluator called this function for the wrong round %d, "+
+				"latest round is %d",
+			round, l.Latest())
+		return
+	}
+
+	return l.GetCreator(cidx, ctype)
+}
+
+func makeLedgerForSimulatorFromLedgerForAPI(ledgerForAPI LedgerForAPI, hdr bookkeeping.BlockHeader) LedgerForSimulator {
+	return &apiSimulatorLedgerConnector{ledgerForAPI, hdr}
+}
+
+// ==============================
+// > Simulator Errors
+// ==============================
+
+// SimulatorError is the base error type for all simulator errors.
+type SimulatorError struct {
+	error
+}
+
+// InvalidTxGroupError occurs when an invalid transaction group was submitted to the simulator.
+type InvalidTxGroupError struct {
+	SimulatorError
+}
+
+// InvalidSignatureError occurs when a transaction has an invalid signature.
+type InvalidSignatureError struct {
+	SimulatorError
+}
+
+// ScopedSimulatorError is a simulator error that has 2 errors, one for internal use and one for
+// displaying publicly. THe external error is useful for API routes/etc.
+type ScopedSimulatorError struct {
+	SimulatorError        // the original error for internal use
+	External       string // the external error for public use
+}
+
+// ==============================
+// > Simulator Helper Methods
+// ==============================
+
+func isInvalidSignatureError(err error) bool {
+	var signatureError verify.SignatureError
+	if errors.As(err, &signatureError) {
+		return true
+	}
+
+	return false
+}
+
+// ==============================
+// > Simulator
+// ==============================
+
+// Simulator is a transaction group simulator for the block evaluator.
+type Simulator struct {
+	ledger LedgerForSimulator
+}
+
+// MakeSimulator creates a new simulator from a ledger.
+func MakeSimulator(ledger LedgerForSimulator) *Simulator {
+	return &Simulator{
+		ledger: ledger,
+	}
+}
+
+// MakeSimulatorFromAPILedger creates a new simulator from an API ledger.
+func MakeSimulatorFromAPILedger(ledgerForAPI LedgerForAPI, hdr bookkeeping.BlockHeader) *Simulator {
+	ledger := makeLedgerForSimulatorFromLedgerForAPI(ledgerForAPI, hdr)
+	return MakeSimulator(ledger)
+}
+
+// checkWellFormed checks that the transaction is well-formed. A failure message is returned if the transaction is not well-formed.
+func (s Simulator) checkWellFormed(txgroup []transactions.SignedTxn) error {
+	hdr, err := s.ledger.BlockHdr(s.ledger.Latest())
+	if err != nil {
+		return ScopedSimulatorError{SimulatorError{fmt.Errorf("please contact us, this shouldn't happen. Current block error: %v", err)}, "current block error"}
```

Comment:
@jdtzmn I think suggesting to contact us is outside the norms.  I found no other similar source code references in go-algorand.

Unless you feel strongly, I leave the following _suggestion_.  I've not yet fully understood error processing, so it's possible there's other candidate error messages here.

```suggestion
		return ScopedSimulatorError{SimulatorError{err}, "current block error"}
```

Reply:
This error message was inspired by [a comment thread Zeph and I had ](https://github.com/algorand/go-algorand/pull/4165#discussion_r922405364)on my Goal box search PR. To summarize our discussion, Indexer uses "contact us" error messages when a bug occurs that shouldn't be possible. Additionally, this error is only logged internally on the Algod side.

I'll go ahead and remove the language, but wanted to include the extra context above 👍

Reply:
@jdtzmn Interesting - thanks for the context. 

Appreciate that you'll bring in the feedback.  For completeness, the added context does _not_ change my view.

Reply:
@michaeldiamant - do you advocate for removing the recently added such message in `goal` in the `feature/avm-box` branch?

More generally, do you think it's worth opening the discussion for Indexer as well?

Reply:
@tzaffi I'm not sure.  Consistent error messages are desirable.  Though it's obviously not our most urgent concern.  I suggest a post-standup discussion to see if we can level set.

* RE:  indexer - At the moment, I don't feel comfortable suggesting changes to indexer.
* RE:  `feature/avm` - I'd make a call based on the proposed verbal.

Reply:
@jdtzmn Resolving as per our verbal with @tzaffi.  We agreed on the following:
* Make the change to this PR as previously agreed.
* Suggest no change in indexer.
* Optionally, if you're up for it @jdtzmn, remove the _please contact us_ language in https://github.com/algorand/go-algorand/pull/4165.